drafts/25-xxx-specifications.txt

```diff
@@ -567,11 +567,121 @@ nd20. The result of evaluating a processor-dependent directive is processor-depe
 5 Expressions allowed in #if and #elif directives
 =================================================
 
-5.1 The 'defined' operator
---------------------------
-
-6 Expression evaluation in #if and #elif directives
-===================================================
+ex05. Controlling expressions are made up of "primary expressions" and
+      operators applied to subexpressions.
+
+ex10. When evaluating a #if or #elif directive, preprocessing
+      evaluates and expands all object-like macros and function-like
+      macro calls to create a token-list of the expression to be
+      evaluated.
+
+ex15. The resulting list of tokens shall be a valid expression
+      comprised of primary expressions and operators as described
+      below.
+
+ex15. Preprocessing computes the integer value of this expressions
+      using the greatest integer range available to the processor to
+      determine the truth or falsity of the controlling expression.
+
+ex20. The evaluation of the expression shall not generate a
+      computational error (such as divide by zero).
+
+ex20. When the expression evaluates to zero, the controlling expression
+      will be considered "false". If the expression evaluates to
+      any non-zero value, the controlling expression will be considered
+      "true".
+
+
+5.1 Primary expressions
+-----------------------
+
+Preprocessing recognizes the following primary expressions in
+controlling expressions.
+
+Since expression evaluation occurs *after* token expansion, there will
+be no object-like macros or function-like macros left to evaluate. All
+instances of ID or ID (args) will all have been replaced with their
+expansions.
+
+We list them here for completeness in describing the illustrative
+syntax of controlling expressions.
+
+| Primary      | Evaluation semantics                        |
+|--------------+---------------------------------------------|
+| ID           | The expansion of the object-like macro ID   |
+| ID (args)    | The expansion of the function-like macro ID |
+| WHOLE_NUMBER | Decimal value of WHOLE_NUMBER               |
+| ( expr )     | Parenthesized expressions                   |
+|--------------+---------------------------------------------|
```

Comment:
Minor aside: C23 6.10.2-13 perversely also allows single-character character constants as "primaries" in conditional expressions, example from C23:
```
#if ’z’ - ’a’ == 25
```
however their exact CPP evaluation semantics are implementation-defined, which means their use is not guaranteed to be portable. I don't believe I've ever seen this bizarre "feature" used in practice. 

I suspect this is some weird legacy holdover in CPP and unless someone provides a strong rationale for their inclusion I think FPP should prohibit character constants in conditional expressions.

Reply:
Agreed, character constants in conditional expressions should be a (pathological) processor-dependent extension.